Libraries/LLM/Cohere.swift

```diff
@@ -144,37 +141,36 @@ public class CohereModelInner: Module {
             mask = mask?.asType(h.dtype)
         }
 
-        var newCache = [(MLXArray, MLXArray)]()
-
         for (i, layer) in layers.enumerated() {
-            var cacheUpdate: (MLXArray, MLXArray)
-            (h, cacheUpdate) = layer(h, mask: mask, cache: cache?[i])
-            newCache.append(cacheUpdate)
+            h = layer(h, mask: mask, cache: cache?[i])
         }
 
-        return (norm(h), newCache)
+        return norm(h)
     }
 }
 
-public class CohereModel: Module, LLMModel {
+public class CohereModel: Module, LLMModel, KVCacheDimensionProvider {
 
     public let vocabularySize: Int
+    public let kvHeads: [Int]
+    public let headDim: IntOrPair
+
     let model: CohereModelInner
     let logitScale: Float
 
     public init(_ args: CohereConfiguration) {
         self.vocabularySize = args.vocabularySize
+        self.kvHeads = (0 ..< args.hiddenLayers).map { _ in args.kvHeads }
+        self.headDim = .init(args.hiddenSize / args.attentionHeads)
         self.model = CohereModelInner(args)
         self.logitScale = args.logitScale
     }
 
-    public func callAsFunction(_ inputs: MLXArray, cache: [(MLXArray, MLXArray)]?) -> (
-        MLXArray, [(MLXArray, MLXArray)]
-    ) {
-        var (out, cache) = model(inputs, cache: cache)
+    public func callAsFunction(_ inputs: MLXArray, cache: [KVCache]?) -> MLXArray {
+        var out = model(inputs, cache: cache)
         out = matmul(out, model.embedTokens.weight.T)
         out = out * self.logitScale
-        return (out, cache)
+        return out
```

Comment:
Every model is changed in this way -- they aren't all identical changes, but 70% are.  The ones that are not identical are just different because of the specifics in the attention layers.